src/blockdiagonal.jl

```diff
@@ -5,16 +5,19 @@
 
 A matrix with matrices on the diagonal, and zeros off the diagonal.
 """
-struct BlockDiagonal{T, V<:AbstractMatrix{T}} <: AbstractMatrix{T}
+struct BlockDiagonal{T, V<:AbstractMatrix{T}, S} <: AbstractMatrix{T}
     blocks::Vector{V}
 
-    function BlockDiagonal{T, V}(blocks::Vector{V}) where {T, V<:AbstractMatrix{T}}
-        return new{T, V}(blocks)
+    function BlockDiagonal{T, V, S}(blocks::Vector{V}) where {T, V<:AbstractMatrix{T}, S}
+        infer_S = all(is_square.(blocks))
+        S == infer_S || throw(ArgumentError("inferred S $infer_S must be equal to S $S"))
```

Comment:
this defines the default constructor meaning we cannot constrcut a `BlockDiagonal` without having to compute `all(is_square.(blocks))` every time... which seems like quite an overhead

we're going to be constructing a new `BlockDiagonal` quite a lot (as the output most mathematical operations)

can we instead have the default constructor expect the `S` arg and have a constructor that allows us passing `S` when we already know it (and similarly have outer-constructors that compute the `S` only when not provided)

Reply:
Yeah, that's a good point. I approached this defensively to prevent bugs by preserving the invariant and didn't give much consideration to efficiency.

Doing some minimal benchmarking it looks like the overhead is small, but not negligible. E.g. consider a simple operation (`-`) where presumably the impact of checking is the relatively the largest*. 

```julia
julia> const blocks = [rand(3, 3) for _ in 1:100];

julia> const bd = BlockDiagonal(blocks);

julia> @btime -bd; # master
  5.222 μs (104 allocations: 13.50 KiB)

julia> @btime -bd; # this PR
  5.416 μs (106 allocations: 13.61 KiB)

julia> @btime all(BlockDiagonals.is_square.(blocks));
  185.885 ns (2 allocations: 112 bytes)
```
so the effect is about 5% slowdown. I am somewhat nervous of allowing the users to make `S` inconsistent with the actual state of the blocks, but since most users will just call `BlockDiagonal(blocks)` maybe that's ok, and allows us to speed up (keep at the same speed really) common math operations by calling the inner constructor directly. What do you think?

*The constructor alone is orders of magnitude slower of course, but that's probably not a relevant comparison since it will always be done in addition to some other operation on blocks.